core/src/cgroup.rs

```diff
@@ -1,238 +1,392 @@
-        let path = PathBuf::from(parent.as_ref()).join(child_name);
-        //trace!("New CGroup: {path:?}");
+    /// Creates a new cgroup as the root of a sub-tree
+    ///
+    /// path must be the path of an already existing cgroup
+    pub fn new_root<P: AsRef<Path>>(path: P, name: &str) -> anyhow::Result<Self> {
+        let path = PathBuf::from(path.as_ref()).join(name);
+
+        // There is no need to check if the path already exists, as create_dir()
+        // will fail under this circumstance.
+        fs::create_dir(&path)?;
+
+        Self::import_root(&path)
+    }
+
+    /// Imports an already existing cgroup as the root of a sub-tree
+    pub fn import_root<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
+        let path = PathBuf::from(path.as_ref());
+
+        if !is_cgroup(&path)? {
+            bail!("{} is not a valid cgroup", path.to_str().unwrap());
+        }
+
+        Ok(CGroup {
+            path: path.to_path_buf(),
+        })
+    }
+
+    /// Creates a sub-cgroup inside this one
+    pub fn new(&self, name: &str) -> anyhow::Result<Self> {
+        Self::new_root(&self.path, name)
+    }
+
+    /// Moves a process to this cgroup
+    pub fn mv(&self, pid: Pid) -> anyhow::Result<()> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
+
+        fs::write(self.path.join("cgroup.procs"), pid.to_string())?;
+        Ok(())
+    }
+
+    /// Returns all PIDs associated with this cgroup
+    pub fn get_pids(&self) -> anyhow::Result<Vec<Pid>> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
 
+        // TODO: "Rustify" this
+        let data = fs::read(self.path.join("cgroup.procs"))?;
+        let mut pids: Vec<Pid> = Vec::new();
+        for line in data.lines() {
+            let line = line?;
+            pids.push(Pid::from_raw(line.parse()?));
+        }
+
+        Ok(pids)
+    }
+
+    /// Checks whether this cgroup is populated
+    pub fn populated(&self) -> anyhow::Result<bool> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
+
+        Ok(!fs::read(self.path.join("cgroup.procs"))?.is_empty())
+    }
+
+    /// Checks whether this cgroup is frozen
+    pub fn frozen(&self) -> anyhow::Result<bool> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
+
+        // We need to check for the existance of cgroup.freeze, because
+        // this file does not exist on the root cgroup.
+        let path = self.path.join("cgroup.freeze");
         if !path.exists() {
-            trace!("Creating Cgroup, {path:?}");
-            std::fs::create_dir(&path).typ(SystemError::CGroup)?;
+            return Ok(false);
+        }
+
+        Ok(fs::read(&path)? == "1\n".as_bytes())
```

Comment:
Instead of `"1\n".to_bytes()` `b"1\n"` can be used. Quotation marks preceded by a `b` result in a byte array